consensus/istanbul/backend/engine.go

```diff
@@ -921,10 +921,73 @@ func (sb *backend) snapshot(chain consensus.ChainReader, number uint64, hash com
 		logger.Trace("Stored voting snapshot to disk", "number", snap.Number, "hash", snap.Hash)
 	}
 
+	sb.regen(chain, headers)
+
 	sb.recents.Add(snap.Hash, snap)
 	return snap, err
 }
 
+// regen commits snapshot data to database
+// regen is triggered if there is any checkpoint block in the `headers`.
+// For each checkpoint block, this function verifies the existence of its snapshot in DB and stores one if missing.
+/*
+ Triggered:
+ |   ^                          ^                          ^                          ^  ...|
+     SI                 SI*(last snapshot)                 SI                         SI
+       			   | header1, .. headerN |
+ Not triggered: (Guaranteed SI* was committed before )
+ |   ^                          ^                          ^                          ^  ...|
+     SI                 SI*(last snapshot)                 SI                         SI
+	                            | header1, .. headerN |
+*/
+func (sb *backend) regen(chain consensus.ChainReader, headers []*types.Header) {
+	// Prevent nested call. Ignore header length one
+	// because it was handled before the `regen` called.
+	defer func() {
+		sb.isRestoringSnapshots.CompareAndSwap(true, false)
+	}()
+	if !sb.isRestoringSnapshots.CompareAndSwap(false, true) || len(headers) <= 1 {
+		return
+	}
```

Comment:
How about this?

```suggestion
	if !sb.isRestoringSnapshots.CompareAndSwap(false, true) || len(headers) <= 1 {
		return
	}
	defer func() {
		sb.isRestoringSnapshots.Store(true)
	}()
```

Reply:
You had a typo `sb.isRestoringSnapshots.Store(true)`? not `false`?

Reply:
oops it's a typo, `false` is right

Reply:
Updated.